### ISSUE TITLE
Bias in early pass training data

### ISSUE THREAD

**Issue by killerducky** [gcp/leela-zero#416]
I think there is a bias problem in the early pass training data. First my evidence of a problem, run this:

```
./leelaz -w 808ac5f35690774587d12f595d6e840801394ee30a26681e4712de1fc2acfb43
play b pass
heatmap
```
The prior for white to pass back and win is 0. I checked some older nets and it was higher but has been steadily going down. It should move towards 100%. Another problem is the prior for black to pass on the first move is holding steady at 3, it should be going down.

I believe there are two causes: 

1) Duplicate games are filtered from training data. This introduces a bias because the correct response to black pass is for white to also pass. Probably there are many of these games but they are filtered. Then all the games were the -n and -m 30 switches caused it to not pass will drown out the good game. Also even the pass/pass games themselves will have different policy targets in them, we should include them all so the NN can average them. Instead we just pick one and depending on what -n did for that one sample we will train on that noise instead of averaging out the noise.

2) Issues #375 and #359 are probably exacerbating the problem, but I think even after that's fixed the other problem will remain. Even in normal games with the current black pass prior on empty boards we will get some portion of black passes due to -n and -m 30.

Edit: Also the value target for black pass should converge on white wins, but because there is only 1 game of white passing back, and many of white just making a move, it will converge to 50/50 instead.

**Comment by bood** [comment]
Good catch!

Another observation is although white prefers corner for the first move now, black still didn't get that yet. Any explanation?

**Comment by jkiliani** [comment]
I don't see it as a problem really. I think what is happening is that the instances of black early passing are disappearing in the training window, and as a consequence, the network is forgetting that there's an easy (cheesy?) answer for white to this due to Tromp-Taylor scoring. This simply means that LeelaZ is no longer considering a pass as a valid move at any point which may result in instant game ending, but as as a generally losing move in early game, like any regular Go player would.

A much stronger Leela Zero may not be able to punish a black first move pass consistently by instant win, but it would punish it by making points and winning the game normally because of that. So this shouldn't really be a problem affecting the playing strength in any way.

**Comment by killerducky** [comment]
> This simply means that LeelaZ is no longer considering a pass as a valid move at any point

The current network policy priors thinks passing is just as good an opening move as playing 4-4, 3-4, or 3-3. I don't call that "pass is not a valid move". If black does pass first, currently it takes more then 3000 playouts for white to realize it can win instantly by passing back.

The problem is easy to fix and I don't see any downside to fixing it. So my opinion is why not fix it?

**Comment by gcp** [comment]
Point (1) is a total non sequitur to me. It literally only affects 1. pass 2. pass which the network hasn't played for a while now.

**Comment by gcp** [comment]
Or do you think the value of passing in the opening position comes solely from games ending there or something? It does not. The network isn't a pure position memory.
If you mean something else I am utterly failing to understand the reasoning in your post.

**Comment by killerducky** [comment]
I agree it doesn't do pure position memory. But I disagree that this isn't causing the NN to evaluate pass on first move differently from pass on other early game moves.

I also disagree that the network hasn't played 1.pass for awhile. With -n -m 30 it does this at almost uniform random probability with any other opening move.

From all.sgf I found 21707 games with 251b7ca8 net.
49 (0.23%) of those games start with black pass. That's very close to uniform random picking pass (1/362=0.28%). Of those, there are two where white passes back and wins, one unix file type and one windows file type (^M at end of line). The other 47 white played some other move. 

I also glanced over games where someone passes in the first line = 10 moves. Almost all cases were it's not the first move and black passes, white passes back. I take this as evidence that the net has learned white can pass back and win in all cases *except* the empty board. That's because there is a larger variety of good training targets since those games are unique, unlike the empty board case.

```
grep 251b7ca8 all.sgf | wc -l
21707
grep -A2 251b7ca8 all.sgf > head_251b7ca8.sgf
grep "^;B\[tt\]" head_251b7ca8.sgf | wc -l
49
grep "\[tt\]" head_251b7ca8.sgf | wc -l
322

```

**Comment by gcp** [comment]
OK, so it does really only affect the starting position? (And the second one proportionally by 1/361th)?

I'm not sure how feasible it is to handle this @roy7. Maybe now that we have seeds, we can allow the same SGF game as long as the seed is different? That would still catch "duplicates".

**Comment by roy7** [comment]
I can make that change if you like. I'll make a new unique hash tied to both instead of just the sgfhash.

If it's only the starting position that is a real problem, you could just add in a few dozen/hundred/whatever example blank board pass/pass results in the training script in every training run. Unless that's considered human knowledge. :)

**Comment by gcp** [comment]
Do please keep the sgfhash too, it's been very useful. Also I think it's the way the training data and SGF are linked in the current schema.

**Comment by killerducky** [comment]
BTW there is no easy way for a user to recreate games since the seed is not stored anywhere. I see you mentioned the sgfhash is useful so I guess putting the seed in the sgf file itself would break that? Maybe I can add it to the debug.txt file and people using autogtp --keep plus --debug could find it there?

**Comment by roy7** [comment]
Putting the random seed in the sgf file won't hurt anything, and in fact, it'd mean the sgfhash for each game would be different if they have different seeds stored.

**Comment by roy7** [comment]
@gcp Should we store seed in an SGF comment so people know what the seed was, and that'll also mean different blank boards with different seeds have different sgfhash and won't be filtered out as duplicates.

**Comment by gcp** [comment]
Clever, I like it.

**Comment by roy7** [comment]
@killerducky @marcocalignano Would one of you toss in a simple PR to add the random seed information to the SGF file as a comment?

**Comment by gcp** [comment]
We should do this in Leela itself BTW, it's useful even without autogtp involved.

**Comment by killerducky** [comment]
I can do it tonight. Should we add the entire command line while we're at it? This can help cases where users submit bugs that are actually bad command line arguments. I can scrub options that might leak user info like directories.

**Comment by gcp** [comment]
That sounds sensible but the scrubbing sounds not so easy to get right?

**Comment by killerducky** [comment]
Something simple like for --logfile just print a generic "-l logfile". For --weights recalculate the hash and print that instead of the user's path.

**Comment by bood** [comment]
maybe json format again, make it easy for script analysis please

2017年12月19日 01:51，"Andy Olsen" <notifications@github.com>写道：

**Comment by gcp** [comment]
>maybe json format again, make it easy for script analysis please

It needs to respect the SGF format.

**Comment by killerducky** [comment]
#426 was merged, so I'm going to close this issue.

In the meantime the network has learned that black passing on first move is very bad, so we are not getting very many examples of 1.pass. When I force a game to do 1.pass, the heatmap still does not give any priority to white pass back. But after 1.pass white can do just about anything and that's enough for it to learn 1.pass is bad.

Maybe if we do a restart we will see different behavior, but it's mostly academic.
